dev/TeachingTip/TeachingTip.cpp

```diff
@@ -0,0 +1,1540 @@
+    TeachingTipTestHooks::NotifyEffectivePlacementChanged(*this);
+    if (IsOpen())
+    {
+        PositionPopup();
+    }
+}
+
+void TeachingTip::OnXCloseButtonClicked(const winrt::IInspectable& sender, const winrt::RoutedEventArgs& args)
+{
+    m_closeButtonClickEventSource(*this, nullptr);
+    m_lastCloseReason = winrt::TeachingTipCloseReason::CloseButton;
+    IsOpen(false);
+}
+
+void TeachingTip::OnCloseButtonClicked(const winrt::IInspectable& sender, const winrt::RoutedEventArgs& args)
+{
+    m_closeButtonClickEventSource(*this, nullptr);
+    m_lastCloseReason = winrt::TeachingTipCloseReason::CloseButton;
+    IsOpen(false);
+}
+
+void TeachingTip::OnActionButtonClicked(const winrt::IInspectable& sender, const winrt::RoutedEventArgs& args)
+{
+    m_actionButtonClickEventSource(*this, nullptr);
+}
+
+void TeachingTip::OnPopupClosed(const winrt::IInspectable& sender, const winrt::IInspectable& args)
+{
+    auto myArgs = winrt::make_self<TeachingTipClosedEventArgs>();
+    if (IsOpen())
+    {
+        m_lastCloseReason = winrt::TeachingTipCloseReason::LightDismiss;
+        IsOpen(false);
+    }
+    myArgs->Reason(m_lastCloseReason);
+    m_closedEventSource(*this, *myArgs);
+}
+
+void TeachingTip::RaiseClosingEvent(const winrt::TeachingTipCloseReason& reason)
+{
+    auto args = winrt::make_self<TeachingTipClosingEventArgs>();
+    args->Reason(reason);
+
+    com_ptr<TeachingTip> strongThis = get_strong();
+    winrt::DeferralCompletedHandler instance{ [strongThis, args]()
+        {
+            strongThis->CheckThread();
+            if (!args->Cancel())
+            {
+                strongThis->ClosePopupWithAnimation();
+            }
+            else
+            {
+                strongThis->IsOpen(true);
+            }
+        }
+    };
+
+    args->SetDeferral(instance);
+
+    args->IncrementDeferralCount();
+    m_closingEventSource(*this, *args);
+    args->DecrementDeferralCount();
+}
+
+void TeachingTip::ClosePopupWithAnimation()
+{
+    if (m_popup && m_popup.get().IsOpen())
+    {
+        StartContractAnimation();
+    }
+}
+
+void TeachingTip::ClosePopup()
+{
+    if (m_popup)
+    {
+        m_popup.get().IsOpen(false);
+    }
+    if (SharedHelpers::IsRS5OrHigher() && m_rootGrid)
+    {
+        //A previous close animation may have left the rootGrid's scale at a very small value and if this teaching tip
+        //is shown again then it's text would be rasturized at this small scale and blown up ~20x. To fix this we have to
+        //reset the scale after the popup has closed so that if the teaching tip is reshown the render pass does not use the
+        //small scale.
+        m_rootGrid.get().Scale(winrt::float3{ 1.0f });
+    }
+}
+
+void TeachingTip::SetBackgroundToDefault()
+{
+    if (IsLightDismissEnabled())
+    {
+        Background(winrt::Application::Current().Resources().Lookup(box_value(s_teachingTipTransientBackgroundBrushName)).as<winrt::Brush>());
```

Comment:
Ideally we never look up resources from code and only do this in Style/VSM to get correct resource lookup logic.